### PR TITLE
Added correctness tests for classification models

### PR DESCRIPTION
**Pull request by fbbradheintz**
This PR provides correctness tests and scriptability checks for all classification models in TorchVision, including Alexnet, Inception, Squeezenet, Modelnet, Googlenet, VGG, MNAS, Densenet, Shufflenet, and Resnet (as well as variations of some of these, such as memory-efficient Densenet).

### CODE REVIEW

File: test/test_models.py [hunk 1]
```diff
@@ -7,17 +7,29 @@
 import unittest
 import traceback
 import random
+import inspect
 
 
-def set_rng_seed(seed):
+EPSILON = 1e-5  # small value for approximate comparisons/assertions
```
Comment (eellison) [inline test/test_models.py]: This is only used once, maybe delete

File: test/test_models.py [hunk 2]
```diff
@@ -40,22 +52,32 @@ def get_available_video_models():
 # they are not yet supported in JIT.
 script_test_models = [
     "deeplabv3_resnet101",
```
Comment (eellison) [inline test/test_models.py]: Why are there only three models here now?

File: test/test_models.py [hunk 3]
```diff
@@ -69,16 +91,255 @@ def check_script(self, model, name):
...
+            tb = traceback.format_exc()
+            actual = False
+            msg = str(e) + str(tb)
+        self.assertEqual(actual, expected, msg)
+
+
+class ClassificationCoverageTester(TestCase):
+
+    # Find all models exposed by torchvision.models factory methods (with assumptions)
+    def get_available_classification_models(self):
```
Comment (eellison) [inline test/test_models.py]: Why are we only testing classification models for this ?

File: test/test_models.py [hunk 4]
```diff
@@ -69,16 +91,255 @@ def check_script(self, model, name):
...
+        model, test_input, test_output = self._test_classification_model(model_callable)
+
+        # above, we perform the standard correctness test against the test fixture, and capture key test params
+        # below, we check that memory efficient/time inefficient DenseNet implementation behaves like the "standard" one
+        me_model = self._get_test_model(model_callable, num_classes=STANDARD_NUM_CLASSES, memory_efficient=True)
+        me_model.load_state_dict(model.state_dict())  # xfer weights over
+        me_output = self._infer_for_test_with(me_model, test_input)
+        test_output.squeeze(0)
+        me_output.squeeze(0)
+        self.assertTrue((test_output - me_output).abs().max() < EPSILON)
```
Comment (eellison) [inline test/test_models.py]: Could we just use `assertNestedTensorObjectsEqual` or use the `assertEqual` that is in @driazati PR here https://github.com/pytorch/vision/pull/1430 ? both of these methods already have a tolerance arg builtin to them. 

Also since EPSILON is only used once i think it would be better to inline the usage here.

File: test/test_models.py [hunk 5]
```diff
@@ -69,16 +91,255 @@ def check_script(self, model, name):
...
+
+# TODO add test for aux_logits arg to factory method
+# TODO add test for transform_input arg to factory method
+class GooglenetTester(ClassificationModelTester):
+    def test_googlenet(self):
+        self._test_classification_model(models.googlenet)
+
+
+class VGGNetTester(ClassificationModelTester):
+    def test_vgg11(self):
```
Comment (eellison) [inline test/test_models.py]: I'm not sure this is an improvement.

File: test/test_models.py [hunk 6]
```diff
@@ -40,22 +52,32 @@ def get_available_video_models():
...
 ]
 
 
 class ModelTester(TestCase):
+
+    # create random tensor with given shape using synced RNG state
+    # caching because these tests take pretty long already (instantiating models and all)
+    TEST_INPUTS = {}
+
+    def _get_test_input(self, shape=STANDARD_INPUT_SHAPE):
```
Comment (eellison) [inline test/test_models.py]: Did you test that this had any noticeable perf improvement?

File: test/test_models.py [hunk 7]
```diff
@@ -69,16 +91,255 @@ def check_script(self, model, name):
...
+            test_methods = test_methods.union(self.get_test_methods_for_class(child))
+        return test_methods
+
+    # Verify that all models exposed by torchvision.models factory methods
+    #    have corresponding test methods
+    # NOTE This does not include some of the extra tests (such as Resnet
+    #    dilation) and says nothing about the correctness of the test, nor
+    #    of the model. It just enforces a naming scheme on the tests, and
+    #    verifies that all models have a corresponding test name.
+    def test_classification_model_coverage(self):
```
Comment (driazati) [inline test/test_models.py]: Why aren't the `torchvision.models` factory methods tested directly?

File: test/test_models.py (hunk 1 above)
Comment (fbbradheintz) [reply test/test_models.py]: It started off being more broadly used, but I refactored most of that away. I still favor keeping a named constant here - either `EPSILON` or `TOLERANCE` - rather than just inserting a magic number with no explanation.

File: test/test_models.py (hunk 2 above)
Comment (fbbradheintz) [reply test/test_models.py]: Because all of the classification models were taken out, because they are all being checked for scriptability.

File: test/test_models.py (hunk 3 above)
Comment (fbbradheintz) [reply test/test_models.py]: The other model types are still using the old, implicit method of test generation.

File: test/test_models.py (hunk 4 above)
Comment (fbbradheintz) [reply test/test_models.py]: I'm happy to switch over to using a more idiomatic assertion call. I'll look into the ones you flagged.

I'm less jazzed about taking away the label, even if it's only used once. Numbers have no semantic content. Not a huge deal in this case, but I prefer a name like `EPSILON` or `TOLERANCE` to a bare constant every time.

File: test/test_models.py (hunk 6 above)
Comment (fbbradheintz) [reply test/test_models.py]: The speed-up is not huge - about a second on a test suite that takes about 200s. It's also not a huge memory hit, since currently only two modestly-sized input tensors are being created. If it becomes an issue, I'm happy to make needed changes.

File: test/test_models.py (hunk 7 above)
Comment (fbbradheintz) [reply test/test_models.py]: The generated tests were implicit and unintentional, and most of the tests were identical, except for a few that were added/tweaked in decision branches. Besides the fact that I was enhancing the test with correctness checks for each model, these generated tests exhibit at least two anti-patterns:

1) Behavior changes that depend on type should be handled by subclassing, not by conditionals.
2) This creates a state where I can add a new model, and a (potentially) passing test gets generated without my ever looking at it. If the passing test is correct, great, I lucked out - but if it's not I have no warning. This is why tests should be explicit and intentional.

In the new code:

1) Test methods are grouped by the models they test. Making a change to ResNet? All the tests for all variants are in one obviously-named class, and you don't have to tease them out of an introspection loop. Differences like the special input shape needed by Inception are handled in the subclass devoted to Inception, not by polluting a conditional branch of omnibus code that tries to cover every variant of every class.
2) If you make a new model, and give it a factory method in `torchvision.models`, `test_classification_model_coverage()` fails the next time you run tests; thus, the dev is nudged to intentionally create a new test that covers the new model/factory method.

So we get more maintainable tests, and lower overhead to having to change tests for new/updated models, and an automatic canary test for new factory methods in `torchvision.models`.

Looking ever further forward, our model offerings are likely to continue to grow, and (I hope) we'll continue building out the test suite, both to cover new models and give better coverage on existing ones. Keeping an increasingly complex test suite for a growing body of code under test in a single class governed by decision branches will turn ugly. The new scheme should have more extensibility and a lower cost of change.

File: test/test_models.py (hunk 5 above)
Comment (fbbradheintz) [reply test/test_models.py]: See my long note below for why I believe it is.

File: test/test_models.py (hunk 7 above)
Comment (eellison) [reply test/test_models.py]: Generally I think it's best to avoid huge refactorings before they're clearly needed. In this case i'm not sure the need has been demonstrated. While the situation you're describing of increasingly complex conditionals seems plausible, it isn't the case today, and I think the updated version here is a substantial increase to the complexity of the code. 
This adds:
     - caching of model inputs per dimensionality
     - introspection of test coverage
     - two separate pathways for testing scriptability
     - two separate pathways for adding model tests. 

I don't think the possibility of someone adding a new model and there being no testing for it is likely. If someone adds a new model, like memory efficient densenet, they're expected to add a new test for it. I don't think this is a difficult standard to uphold as a contributor or reviewer.

File: test/test_models.py (hunk 3 above)
Comment (eellison) [reply test/test_models.py]: I don't think there should be two separate pathways at one time.

File: test/test_models.py [hunk 8]
```diff
@@ -7,17 +7,29 @@
...
 import traceback
 import random
+import inspect
 
 
-def set_rng_seed(seed):
+EPSILON = 1e-5  # small value for approximate comparisons/assertions
+STANDARD_NUM_CLASSES = 50
+STANDARD_INPUT_SHAPE = (1, 3, 224, 224)
+STANDARD_SEED = 1729  # https://fburl.com/3i5wkg9p
```
[the remark this answers is not in the document]
Comment (fbbradheintz) [reply test/test_models.py]: doh - good catch

File: test/test_models.py (hunk 3 above)
Comment (fbbradheintz) [reply test/test_models.py]: I agree, but my brief was to cover the classification models. Also, I think the four other major classes of model should probably get broken out into separate files.

File: test/test_models.py (hunk 4 above)
Comment (fbbradheintz) [reply test/test_models.py]: Cool - named arguments serve the semantic purpose. I'll tweak this.

File: test/test_models.py [hunk 9]
```diff
@@ -7,17 +7,29 @@
...
+
+
+def set_rng_seed(seed=STANDARD_SEED):
```
[the remark this answers is not in the document]
Comment (fbbradheintz) [reply test/test_models.py]: Let me know when that gets merged, happy to use it.

File: test/test_models.py (hunk 7 above)
Comment (fbbradheintz) [reply test/test_models.py]: This refactoring *was* needed in order to work with the tests in isolation. Those implicit, unintentional tests were a PITA to isolate and change, which limits their utility. Also, the correctness checks themselves are an example of the building complexity - they are sometimes slightly different for different models, e.g. checking that the memory efficient Densenet implementation showed the same output as the vanilla version. To your other points:

* I'm not married to the test input caching, and if there's a demonstrable issue that I'm not aware of (as opposed to merely a prospective one), I'm happy to change it.
* The introspection was already happening. I only included it so we wouldn't lose the sense of security people seem to get from the implicit/unintentional tests being generated over the factory methods.
* If by "two separate pathways" you mean the difference between classification and other model types, that's true. My brief only covered Classification models. I think the other model types should be broken out into their own files (as their testing needs will likely differ), and updated to dump the opaque generated tests.
* Yes, there is more code, but it's also modularized in a way that explicitly maps the tests to the code under test, removes conditionals in favor of OO, and makes it trivially obvious to find where to make a change or set a breakpoint for a particular model with a text search. "Less code" is not, in every situation, the best variable to optimize for.

Your last point - about test being enforced by process - is orthogonal to the point I made above. If someone adds a model, never looks at the (implicit, unintentional) tests, and the automated tests all show green, the likelihood is against either the dev or a reviewer going deeper. The generated tests leave that door open. The new tests do nothing if the dev codes their tests first, and gives a nudge if they don't.

**Comment by fbbradheintz** [comment]
Per conversation with @fmassa, I'm going merging this.
